crates/walrus-indexer/src/async_task_manager.rs

```diff
@@ -0,0 +1,705 @@
+// Copyright (c) Walrus Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! Async task manager built on top of PersistentQueue.
+
+#![allow(dead_code)]
+
+use std::sync::Arc;
+
+use anyhow::Result;
+use tokio::{
+    sync::{Mutex, Semaphore, mpsc},
+    task::JoinHandle,
+    time::Duration,
+};
+use tokio_util::sync::CancellationToken;
+
+use crate::{
+    AsyncTask,
+    AsyncTaskStore,
+    TaskExecutor,
+    async_task_sorter::{AsyncTaskSorter, AsyncTaskSorterConfig},
+};
+
+/// Handles retry queue processing.
+/// It scans the retry queue, and spawns workers for any retry tasks it finds.
+/// The process is repeated as long as there exists retry tasks in the queue.
+struct RetryQueueHandler<T, S, E>
+where
+    T: AsyncTask,
+    S: AsyncTaskStore<T>,
+    E: TaskExecutor<T>,
+{
+    store: Arc<S>,
+    executor: Arc<E>,
+    config: AsyncTaskManagerConfig,
+    retry_notification_receiver: mpsc::UnboundedReceiver<()>,
+    retry_notification_sender: mpsc::UnboundedSender<()>,
+    shutdown_token: CancellationToken,
+    /// Synchronization mutex for task_handle and retry_notified
+    sync_mutex: Arc<Mutex<()>>,
+    task_handle: Option<JoinHandle<()>>,
+    /// A hint whether there are retry tasks to process.
+    retry_notified: bool,
```

Comment:
remark: is the mutex mainly for guarding access to task_handle and task_handle? Should these two variables inside the mutex?